test/common/common/mutex_tracer_test.cc

```diff
@@ -0,0 +1,118 @@
+#include <chrono>
+#include <thread>
+
+#include "common/common/mutex_tracer.h"
+
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class MutexTracerTest : public testing::Test {
+protected:
+  void SetUp() override {
+    tracer_ = MutexTracer::getOrCreateTracer();
+    tracer_->reset();
+  }
+
+  // Since MutexTracer::contentionHook is a private method, MutexTracerTest is a friend class.
+  void sendWaitCyclesToContentionHook(int64_t wait_cycles) {
+    tracer_->contentionHook(nullptr, nullptr, wait_cycles);
+  }
+
+  absl::Mutex mu_;
+  MutexTracer* tracer_;
+};
+
+// Call the contention hook manually.
+TEST_F(MutexTracerTest, AddN) {
+  EXPECT_EQ(tracer_->numContentions(), 0);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 0);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 0);
+
+  sendWaitCyclesToContentionHook(2);
+
+  EXPECT_EQ(tracer_->numContentions(), 1);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 2);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 2);
+
+  sendWaitCyclesToContentionHook(3);
+
+  EXPECT_EQ(tracer_->numContentions(), 2);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 3);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 5);
+
+  sendWaitCyclesToContentionHook(0);
+
+  EXPECT_EQ(tracer_->numContentions(), 3);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 0);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 5);
+}
+
+// Call the contention hook in a real contention scenario.
+TEST_F(MutexTracerTest, OneThreadNoContention) {
+  // Regular operation doesn't cause contention.
+  mu_.Lock();
+  mu_.Unlock();
+
+  EXPECT_EQ(tracer_->numContentions(), 0);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 0);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 0);
+}
+
+TEST_F(MutexTracerTest, TryLockNoContention) {
+  // TryLocks don't cause contention.
+  mu_.Lock();
+  EXPECT_FALSE(mu_.TryLock());
+  mu_.Unlock();
+
+  EXPECT_EQ(tracer_->numContentions(), 0);
+  EXPECT_EQ(tracer_->currentWaitCycles(), 0);
+  EXPECT_EQ(tracer_->lifetimeWaitCycles(), 0);
+}
+
+void holdMutexUntil(absl::Mutex* mu, absl::Notification* notif_unlock) {
+  mu->Lock();
+  notif_unlock->WaitForNotification();
```

Comment:
This can't work because (as I have observed experimentally), mutex contention stats are not collected / observable by tracer_->numContentions() until the thread is rejoined.

Reply:
(I think) this doesn't work because absl mutex contention doesn't register that there was contention until the mutex was released by its initial holder. I have observed this experimentally but am not totally sure. Will report back.

Reply:
If that's the case, then I recommend dropping the Notification objects, and just spinning in the thread-functions like:
```
  while (content-count unchanged) {
    test_time_.sleep(std::chrono::miliseconds(1));
    LockGuard lock(mu_);
    test_time_.sleep(std::chrono::miliseconds(9));
  }
```
With two threads holding the lock 90% of the time it shouldn't take too long for them to contend. I don't think it will be flaky, but it would be worth testing with `--runs_per_test=1000` just to get confidence.

Reply:
...and in the test method you really just need to start up both threads and join them.

Reply:
Done in ac7f65c.